crates/router/src/core/payments/operations/payment_response.rs

```diff
@@ -1243,6 +1243,54 @@ impl<F: Clone> PostUpdateTracker<F, PaymentData<F>, types::CompleteAuthorizeData
     }
 }
 
+async fn update_payment_method_type_for_cards<F: Clone>(
+    state: &SessionState,
+    storage_scheme: enums::MerchantStorageScheme,
+    payment_data: &mut PaymentData<F>,
+) -> RouterResult<()> {
+    // Parse Value to AdditionalPaymentData
+    let additional_payment_method_data: Option<api_models::payments::AdditionalPaymentData> =
+        payment_data
+            .get_payment_attempt()
+            .payment_method_data
+            .clone()
+            .map(|data| data.parse_value("payment_method_data"))
+            .transpose()
+            .change_context(errors::ApiErrorResponse::InvalidDataValue {
+                field_name: "payment_method_data",
+            })?;
+
+    // Fetch card_type from this struct for cards
+    let payment_method_type = match additional_payment_method_data {
+        Some(api_models::payments::AdditionalPaymentData::Card(card)) => card
+            .card_type
+            .map(|card_type_str| enums::PaymentMethodType::from_str(&card_type_str.to_lowercase()))
+            .transpose()
+            .change_context(errors::ApiErrorResponse::InternalServerError)?,
+        _ => None,
+    };
+
+    // If payment_method_type in PaymentAttempt does not match the one in payment_method_data
+    // update payment_method_type in PaymentAttempt table
+    if payment_data.payment_attempt.payment_method_type.as_ref() != payment_method_type.as_ref() {
```

Comment:
If above initialised `payment_method_type` is `None` and `payment_data.payment_attempt.payment_method_type` already has some value in it, wouldn't attempt's payment_method_type be updated with `None`?

Reply:
That does not occur, but I think we can make it more specific!